scylla-cql/src/frame/value.rs

```diff
@@ -879,64 +878,66 @@ impl<'b> ValueList for Cow<'b, SerializedValues> {
 /// The underlying iterator will always be cloned at least once, once to compute the length if it can't be known
 /// in advance, and be re-cloned at every retry.
 /// It is consequently expected that the provided iterator is cheap to clone (e.g. `slice.iter().map(...)`).
-pub struct BatchValuesFromIter<IT> {
+pub struct BatchValuesFromIter<'a, IT> {
     it: IT,
+    _spooky: std::marker::PhantomData<&'a ()>,
 }
 
-impl<'a, IT, VL> BatchValuesFromIter<IT>
+impl<'a, IT, VL> BatchValuesFromIter<'a, IT>
 where
     IT: Iterator<Item = &'a VL> + Clone,
     VL: ValueList + 'a,
 {
     pub fn new(into_iter: impl IntoIterator<IntoIter = IT>) -> Self {
         Self {
             it: into_iter.into_iter(),
+            _spooky: std::marker::PhantomData,
         }
     }
 }
 
-impl<IT> From<IT> for BatchValuesFromIter<IT>
+impl<'a, IT, VL> From<IT> for BatchValuesFromIter<'a, IT>
 where
-    IT: Iterator + Clone,
-    IT::Item: ValueList,
+    IT: Iterator<Item = &'a VL> + Clone,
```

Comment:
> Doesn't this make the bound more restrictive than before?

No that was actually a mistake before: the implementation of `From` was not restrictive enough compared to the implementation of `BatchValues` for `BatchValuesFromIter`:
https://github.com/scylladb/scylla-rust-driver/blob/d112dc2cf1c770cd2609e2670cc61f04dd05d7b6/scylla-cql/src/frame/value.rs#L922-L926

It seems that I noticed it as I was porting to actual GATs but forgot to backport the fix.

> Could this restriction be easily relaxed?

That is something I initially attempted, but ended up moving away from it because the signature of `ValueList` is such that the lifetime `'a` of the `SerializedResult<'a>` is bound to the lifetime of the `&'a impl ValueList`, which means the `SerializedResult<'a>` can only live as long as the `ValueList` is in scope. Now if the iterator yields `ValueList`s directly, that means it has to be stored in memory / unmoved for the `SerializedResult<'a>` to stay alive, which would require storing that in `SerializedValuesIterator` and not reusing that `SerializedValuesIterator` for the time the `SerializedResult<'a>` is alive, which would prevent users of `BatchValues` from having several `SerializedResult<'a>` live at a time (which currently is used by the `batch` code).

Reply:
I see, thanks for the explanation. Allowing this use case would probably require a comprehensive redesign of the serialization framework which is out of scope of this PR.